File: FreePascal/NBody.pas
(* The Computer Language Benchmarks Game
 * http://shootout.alioth.debian.org/
 *
 * Based on nbody.java and adapted based on the SOM version.
 * Copyright (c) 2023 Rochus Keller <me@rochus-keller.ch> (for FreePascal migration)
 *)
 
unit NBody;
interface
uses Benchmark;

type
	PNBody = ^TNBody;
	TNBody = object(TBenchmark)
		    constructor init;
		    destructor deinit; virtual;
		    function benchmark(): longint; virtual;
		    function verifyResult(result: longint): boolean; virtual;
		    function innerBenchmarkLoop(innerIterations: longint): boolean; virtual;
		end;
		
implementation
	
    const
        PI = double(3.141592653589793);
        SOLAR_MASS = double(4.0 * PI * PI);
        DAYS_PER_YER = double(365.24);
     	numberOfBodies = 5;
       
    type
        TBody = object
                    x,y,z,vx,vy,vz,mass: double;
					procedure offsetMomentum(px, py, pz: double);
                end;
        NBodySystem = object 
        			bodies: array [0..numberOfBodies-1] of TBody;
        			constructor init();
        			procedure advance(dt: double);
        			function energy(): double;
        		end;

	procedure TBody.offsetMomentum(px, py, pz: double);
	begin
	   self.vx := 0.0 - (px / SOLAR_MASS);
	   self.vy := 0.0 - (py / SOLAR_MASS);
	   self.vz := 0.0 - (pz / SOLAR_MASS);
	end;

	procedure NBodySystem.advance(dt: double);
		var i,j: longint;
		    dx,dy,dz,dSquared,distance,mag: double;
		    iBody,jBody: TBody;
	begin
	   for i := 0 to numberOfBodies-1 do
	   begin
		  iBody := self.bodies[i];
		  for j := i + 1 to numberOfBodies-1 do
		  begin
		     jBody := self.bodies[j];
		     dx := iBody.x - jBody.x;
		     dy := iBody.y - jBody.y;
		     dz := iBody.z - jBody.z;

		     dSquared := dx * dx + dy * dy + dz * dz;
		     distance := sqrt(dSquared);
		     mag := dt / (dSquared * distance);

		     iBody.vx := iBody.vx - (dx * jBody.mass * mag);
		     iBody.vy := iBody.vy - (dy * jBody.mass * mag);
		     iBody.vz := iBody.vz - (dz * jBody.mass * mag);

		     jBody.vx := jBody.vx + (dx * iBody.mass * mag);
		     jBody.vy := jBody.vy + (dy * iBody.mass * mag);
		     jBody.vz := jBody.vz + (dz * iBody.mass * mag);
		  end;
		end;

	   	for i := 0 to numberOfBodies-1 do
	   	begin
		  iBody := self.bodies[i];
		  iBody.x := iBody.x + dt * iBody.vx;
		  iBody.y := iBody.y + dt * iBody.vy;
		  iBody.z := iBody.z + dt * iBody.vz;
	    end;
	end;

	function NBodySystem.energy(): double;
	   var dx,dy,dz,distance,e: double;
		   iBody,jBody: TBody;
		   i,j: longint;
	begin
	   e := 0.0;

	  for i := 0 to numberOfBodies-1 do
	  begin
		  iBody := self.bodies[i];
		  e := e + 0.5 * iBody.mass
		      * (iBody.vx * iBody.vx +
		         iBody.vy * iBody.vy +
		         iBody.vz * iBody.vz);

		  for j := i + 1 to numberOfBodies-1 do
		  begin
		     jBody := self.bodies[j];
		     dx := iBody.x - jBody.x;
		     dy := iBody.y - jBody.y;
		     dz := iBody.z - jBody.z;

		    distance := sqrt( dx * dx + dy * dy + dz * dz );
		    e := e - (iBody.mass * jBody.mass) / distance;
		  end;
	   end;
	   energy := e;
	end;

    function createBody(x,y,z,vx,vy,vz,mass: double): TBody;
        var b: TBody;
    begin
        b.x := x;
        b.y := y;
        b.z := z;
        b.vx := vx * DAYS_PER_YER;
        b.vy := vy * DAYS_PER_YER;
        b.vz := vz * DAYS_PER_YER;
        b.mass := mass * SOLAR_MASS;
        exit(b);
    end;

    function jupiter(): TBody;
    begin
    	exit(createBody(
         4.84143144246472090e+00,
        -1.16032004402742839e+00,
        -1.03622044471123109e-01,
         1.66007664274403694e-03,
         7.69901118419740425e-03,
        -6.90460016972063023e-05,
         9.54791938424326609e-04));
    end;

    function saturn(): TBody;
    begin
    	exit(createBody(
         8.34336671824457987e+00,
         4.12479856412430479e+00,
        -4.03523417114321381e-01,
        -2.76742510726862411e-03,
         4.99852801234917238e-03,
         2.30417297573763929e-05,
         2.85885980666130812e-04)); 
    end;

    function uranus(): TBody;
    begin
    	exit(createBody(
         1.28943695621391310e+01,
        -1.51111514016986312e+01,
        -2.23307578892655734e-01,
         2.96460137564761618e-03,
         2.37847173959480950e-03,
        -2.96589568540237556e-05,
         4.36624404335156298e-05));
    end;

    function neptune(): TBody;
    begin
    	exit(createBody(
         1.53796971148509165e+01,
        -2.59193146099879641e+01,
         1.79258772950371181e-01,
         2.68067772490389322e-03,
         1.62824170038242295e-03,
        -9.51592254519715870e-05,
         5.15138902046611451e-05));
    end;

    function sun(): TBody;
    begin
    	exit(createBody(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0));
    end;
    
    constructor NBodySystem.init();
        var i: longint;
            px,py,pz: double;
    begin        
        self.bodies[0] := sun();
        self.bodies[1] := jupiter();
        self.bodies[2] := saturn();
        self.bodies[3] := uranus();
        self.bodies[4] := neptune();
        
        px := 0.0; py := 0.0; pz := 0.0;
        for i := 0 to numberOfBodies-1 do
        begin
            px := px + self.bodies[i].vx * self.bodies[i].mass;
            py := py + self.bodies[i].vy * self.bodies[i].mass;
            pz := pz + self.bodies[i].vz * self.bodies[i].mass;
        end;

        self.bodies[0].offsetMomentum(px, py, pz);
    end;
    
    constructor TNBody.init;
    begin
    	inherited init;
    end;
    
    destructor TNBody.deinit;
    begin
    	inherited deinit;
    end;
    
    function TNBody.benchmark(): longint;
    begin
    	halt;
    	exit(0);
    end;
    
    function TNBody.verifyResult(result: longint): boolean;
    begin
    	halt;
    	exit(false);
    end;

	function verifyResult2(result: double; innerIterations: longint):boolean;
	const epsilon = 0.00005; // TODO 0.00000000000000005; // 5e-17
		// looks like the precision is only single, not double; somewhere precision gets lost
	begin
		if innerIterations = 250000 then exit(abs(result) - 0.1690859889909308 < epsilon);
		if innerIterations = 1 then exit(abs(result) - 0.16907495402506745 < epsilon);
													 //0.16907516382852447

		// Checkstyle: stop
		Write('No verification result for ');
		Write(innerIterations);
		WriteLn(' found');
		Write('Result is: ');
		WriteLn(result);
		// Checkstyle: resume
		exit(false);
	end;

    function TNBody.innerBenchmarkLoop(innerIterations: longint): boolean; 
	var system: NBodySystem;
        i: longint;
    begin
        system.init();
        for i := 0 to innerIterations-1 do
        	system.advance(0.01);
        exit(verifyResult2(system.energy(), innerIterations));
    end;

end.

File: C/NBody.c
/* The Computer Language Benchmarks Game
 * http://shootout.alioth.debian.org/
 *
 * Based on nbody.java and adapted based on the SOM version.
 * Copyright (c) 2024 Rochus Keller <me@rochus-keller.ch> (for C99 migration)
 */

#include "NBody.h"
#include <math.h>
#include <stdlib.h>
#include <stdio.h>
#include <assert.h>

#define PI  3.141592653589793
#define SOLAR_MASS (4 * PI * PI)
#define DAYS_PER_YER 365.24

typedef struct Body {
    double x;
    double y;
    double z;
    double vx;
    double vy;
    double vz;
    double mass;
} Body;

static double Body_getX(Body* me) { return me->x; }
static double Body_getY(Body* me) { return me->y; }
static double Body_getZ(Body* me) { return me->z; }

static double Body_getVX(Body* me) { return me->vx; }
static double Body_getVY(Body* me) { return me->vy; }
static double Body_getVZ(Body* me) { return me->vz; }

static double Body_getMass(Body* me) { return me->mass; }

static void Body_setX(Body* me, double x) { me->x = x; }
static void Body_setY(Body* me, double y) { me->y = y; }
static void Body_setZ(Body* me, double z) { me->z = z; }

static void Body_setVX(Body* me, double vx) { me->vx = vx; }
static void Body_setVY(Body* me, double vy) { me->vy = vy; }
static void Body_setVZ(Body* me, double vz) { me->vz = vz; }

static void Body_offsetMomentum(Body* me, double px, double py, double pz) {
    me->vx = 0.0 - (px / SOLAR_MASS);
    me->vy = 0.0 - (py / SOLAR_MASS);
    me->vz = 0.0 - (pz / SOLAR_MASS);
}

Body Body_create(double x, double y, double z,
     double vx, double vy, double vz, double mass) {
    Body res;
    Body* me = &res;
    me->x = x;
    me->y = y;
    me->z = z;
    me->vx = vx * DAYS_PER_YER;
    me->vy = vy * DAYS_PER_YER;
    me->vz = vz * DAYS_PER_YER;
    me->mass = mass * SOLAR_MASS;
    return res;
}

static Body Body_jupiter() {
    return Body_create(
                4.84143144246472090e+00,
                -1.16032004402742839e+00,
                -1.03622044471123109e-01,
                1.66007664274403694e-03,
                7.69901118419740425e-03,
                -6.90460016972063023e-05,
                9.54791938424326609e-04);
}

static Body Body_saturn() {
    return Body_create(
                8.34336671824457987e+00,
                4.12479856412430479e+00,
                -4.03523417114321381e-01,
                -2.76742510726862411e-03,
                4.99852801234917238e-03,
                2.30417297573763929e-05,
                2.85885980666130812e-04);
}

static Body Body_uranus() {
    return Body_create(
                1.28943695621391310e+01,
                -1.51111514016986312e+01,
                -2.23307578892655734e-01,
                2.96460137564761618e-03,
                2.37847173959480950e-03,
                -2.96589568540237556e-05,
                4.36624404335156298e-05);
}

static Body Body_neptune() {
    return Body_create(
                1.53796971148509165e+01,
                -2.59193146099879641e+01,
                1.79258772950371181e-01,
                2.68067772490389322e-03,
                1.62824170038242295e-03,
                -9.51592254519715870e-05,
                5.15138902046611451e-05);
}

static Body Body_sun() {
    return Body_create(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0);
}

enum { Count = 5 };

typedef struct NBodySystem {
    Body bodies[Count];
} NBodySystem;

static void NBodySystem_init(NBodySystem* me) {
    me->bodies[0] = Body_sun();
    me->bodies[1] = Body_jupiter();
    me->bodies[2] = Body_saturn();
    me->bodies[3] = Body_uranus();
    me->bodies[4] = Body_neptune();

    double px = 0.0;
    double py = 0.0;
    double pz = 0.0;

    for ( int i = 0; i < Count; i++ ) {
        Body* b = &me->bodies[i];
        px += Body_getVX(b) * Body_getMass(b);
        py += Body_getVY(b) * Body_getMass(b);
        pz += Body_getVZ(b) * Body_getMass(b);
    }

    Body_offsetMomentum(&me->bodies[0], px, py, pz);
}

static void NBodySystem_advance(NBodySystem* me, double dt) {

    for (int i = 0; i < Count; ++i) {
        Body* iBody = &me->bodies[i];

        for (int j = i + 1; j < Count; ++j) {
            Body* jBody = &me->bodies[j];
            const double dx = Body_getX(iBody) - Body_getX(jBody);
            const double dy = Body_getY(iBody) - Body_getY(jBody);
            const double dz = Body_getZ(iBody) - Body_getZ(jBody);

            const double dSquared = dx * dx + dy * dy + dz * dz;
            const double distance = sqrt(dSquared);
            const double mag = dt / (dSquared * distance);

            Body_setVX(iBody, Body_getVX(iBody) - (dx * Body_getMass(jBody) * mag));
            Body_setVY(iBody, Body_getVY(iBody) - (dy * Body_getMass(jBody) * mag));
            Body_setVZ(iBody, Body_getVZ(iBody) - (dz * Body_getMass(jBody) * mag));

            Body_setVX(jBody, Body_getVX(jBody) + (dx * Body_getMass(iBody) * mag));
            Body_setVY(jBody, Body_getVY(jBody) + (dy * Body_getMass(iBody) * mag));
            Body_setVZ(jBody, Body_getVZ(jBody) + (dz * Body_getMass(iBody) * mag));
        }
    }

    for (int i = 0; i < Count; i++) {
        Body* body = &me->bodies[i];
        Body_setX(body, Body_getX(body) + dt * Body_getVX(body));
        Body_setY(body, Body_getY(body) + dt * Body_getVY(body));
        Body_setZ(body, Body_getZ(body) + dt * Body_getVZ(body));
    }
}

static double NBodySystem_energy(NBodySystem* me) {
    double e = 0.0;

    for (int i = 0; i < Count; ++i) {
        Body* iBody = &me->bodies[i];
        e += 0.5 * Body_getMass(iBody)
                * (Body_getVX(iBody) * Body_getVX(iBody) +
                   Body_getVY(iBody) * Body_getVY(iBody) +
                   Body_getVZ(iBody) * Body_getVZ(iBody));

        for (int j = i + 1; j < Count; ++j) {
            Body* jBody = &me->bodies[j];
            const double dx = Body_getX(iBody) - Body_getX(jBody);
            const double dy = Body_getY(iBody) - Body_getY(jBody);
            const double dz = Body_getZ(iBody) - Body_getZ(jBody);

            const double distance = sqrt(dx * dx + dy * dy + dz * dz);
            e -= (Body_getMass(iBody) * Body_getMass(jBody)) / distance;
        }
    }
    return e;
}

static bool verifyResult2(double result, int innerIterations)
{
    const double epsilon = 0.00000000000000005; // 5e-17
    if (innerIterations == 250000) {
        return abs(result) - 0.1690859889909308 < epsilon;
    }
    if (innerIterations == 1) {
        return abs(result) - 0.16907495402506745 < epsilon;
    }

    // Checkstyle: stop
    fprintf(stderr,"No verification result for %d found\n", innerIterations);
    fprintf(stderr,"Result is: %f\n", result);
    // Checkstyle: resume
    return false;
}

static bool innerBenchmarkLoop(Benchmark* me, int innerIterations)
{
    NBodySystem system;
    NBodySystem_init(&system);
    for (int i = 0; i < innerIterations; i++) {
        NBodySystem_advance(&system,0.01);
    }

    return verifyResult2(NBodySystem_energy(&system), innerIterations);
}

static int benchmark(Benchmark* me) {
    assert(0); // "Should never be reached"
}

static bool verifyResult(Benchmark* me, int result) {
    assert(0); // "Should never be reached"
}


Benchmark*NBody_create()
{
    Benchmark* bench = (Benchmark*)malloc(sizeof(Benchmark));
    bench->benchmark = benchmark;
    bench->verifyResult = verifyResult;
    bench->dispose = 0;
    bench->innerBenchmarkLoop = innerBenchmarkLoop;
    return bench;
}
